add task solution

Andriana1112 opened the pull request.

Commit:
add task solution

mateacademy-ai-mentor commented on src/models/User.js:
```diff
@@ -0,0 +1,212 @@
+    this.email = newEmail;
+
+    return true;
+  }
```
The `updateEmail` method does not require the current password for authorization, nor does it require confirmation of the new email (i.e., double entry or confirmation step). Additionally, it does not send a notification to the old email about the change, which is required by the checklist. Consider updating this method to enforce these requirements.

mateacademy-ai-mentor commented on src/routes/auth.js:
```diff
@@ -0,0 +1,310 @@
+);
+
+// Вихід
+router.get('/logout', (req, res) => {
```
The `/logout` route does not use the `isAuthenticated` middleware, so it is accessible to non-authenticated users. According to the requirements, logout must be accessible only to authenticated users. Add the `isAuthenticated` middleware to this route.

mateacademy-ai-mentor commented on src/routes/auth.js:
```diff
@@ -0,0 +1,310 @@
+      }
+    } catch (error) {
+      req.flash('error', error.message);
+      res.redirect('/auth/forgot-password');
```
In the password reset request handler, if the email is not found, an error is thrown and the user is redirected back with an error message. This could leak information about which emails are registered. The checklist suggests (optionally) always showing the same confirmation page regardless of whether the email exists, to avoid information leakage. Consider handling this as a best practice.

mateacademy-ai-mentor commented on src/routes/profile.js:
```diff
@@ -0,0 +1,170 @@
+      .isEmail()
+      .normalizeEmail()
+      .withMessage('Введіть коректну нову електронну пошту'),
+  ],
```
The email change flow does not require confirmation of the new email (i.e., double entry or confirmation step). According to the requirements, you should add a `confirmNewEmail` field and validate that it matches `newEmail` before updating the email.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

**Decision: Request Changes**

You’ve implemented the majority of the auth and profile features cleanly and in line with the checklist. However, there are a few critical blockers that violate explicit requirements and need to be addressed before approval.

**Critical issues to fix (blockers):**
1. Logout route access control
   - Location: `src/routes/auth.js` (e.g., around the `/logout` route definition; likely near lines where other auth routes are defined)
   - Issue: `/logout` does not use `isAuthenticated`, making it accessible to unauthenticated users.
   - Why it’s wrong: The checklist requires logout to be accessible only to authenticated users.
   - How to fix: Add the middleware to the route, for example: `router.post('/logout', isAuthenticated, ...)` (or `get` depending on your method), and keep the redirect behavior unchanged.

2. Email change flow requires confirmation and current password
   - Locations:
     - `src/routes/profile.js` (email change handler; likely where `newEmail` is validated)
     - `src/models/User.js` (`updateEmail` method)
   - Issues:
     - No `confirmNewEmail` field or matching validation against `newEmail`.
     - `updateEmail` does not verify the user’s current password before changing the email.
     - The checklist explicitly requires both confirmation of the new email and current password verification.
   - Why it’s wrong: This weakens both security and UX for a sensitive change.
   - How to fix:
     - In `profile.js`: require `confirmNewEmail` in the form and validate `newEmail === confirmNewEmail` before proceeding; return a validation error if they differ.
     - In `User.updateEmail(...)`: accept and verify the current password (e.g., compare with stored hash) before applying the email change. If you prefer to keep password checks in routes, ensure it’s strictly enforced there before calling the model.
     - Ensure a notification to the old email continues to be sent after a successful change (this appears already implemented in `emailService`).

**Best-practice improvement (non-blocker):**
- Password reset request information leakage
  - Location: `src/routes/auth.js` (forgot-password handler)
  - Suggestion: Always show the same success/confirmation response whether or not the email exists to avoid user enumeration.

You’re very close—middleware usage and overall structure are solid. Addressing the logout protection and tightening the email change flow (confirm + current password) will bring you to full compliance. Keep up the great work!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

Commit:
fix: resolve critical security issues

- Add isAuthenticated middleware to /logout route
- Add email confirmation field for email change
- Implement email confirmation validation
- Remove password parameter from updateEmail method
- Improve password reset security (prevent user enumeration)
- Fix all linter errors

mateacademy-ai-mentor commented on src/models/User.js:
```diff
@@ -0,0 +1,212 @@
+  }
+
+  // Оновлення email
+  updateEmail(newEmail) {
```
The `updateEmail` method does not require verification of the user's current password before changing the email, nor does it require confirmation of the new email (i.e., `confirmNewEmail`). According to the requirements, changing email must require the user's current password and confirmation of the new email. Ensure these checks are enforced in the route handler before calling this method, or implement them here if preferred.

mateacademy-ai-mentor reviewed (approved):
### Review Summary

**Decision:** Approve — core requirements are met.

- You’ve implemented all required features across auth, sessions, access control, and profile management. The email change flow in `src/routes/profile.js` now correctly requires the current password and `confirmNewEmail`, addressing the critical concern.
- In `src/models/User.js`, `updateEmail` itself doesn’t enforce password or confirmation checks, but you’ve correctly handled these validations in the route before calling the method, which satisfies the requirement. If you keep this contract, consider adding a brief comment near `updateEmail` to document that the route must verify the current password and new email confirmation before invoking it.
- Everything else looks solid: validation, token handling, error/flash messages, and overall structure are well done.

Nice work closing the gaps and delivering a secure, user-friendly implementation!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)
